File: 0148_sort_list/sort_list.java
public class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    // Time: O(n * log(n))
    // Space: O(1)
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode mid = getMid(head);
        ListNode left = sortList(head);
        ListNode right = sortList(mid);
        return merge(left, right);

    }

    private ListNode getMid(ListNode head) {
        ListNode midPrev = null;
        while (head != null && head.next != null) {
            midPrev = (midPrev == null) ? head : midPrev.next;
            head = head.next.next;
        }
        ListNode mid = midPrev.next;
        midPrev.next = null;
        return mid;
    }

    private ListNode merge(ListNode first, ListNode sec) {
        ListNode sentinel = new ListNode();
        ListNode tail = sentinel;
        while (first != null && sec != null) {
            if (first.val < sec.val) {
                tail.next = first;
                first = first.next;
            } else {
                tail.next = sec;
                sec = sec.next;
            }
            tail = tail.next;
        }
        tail.next = (first != null) ? first : sec;
        return sentinel.next;
    }
}

File: 0148_sort_list/sort_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def sortList(head: ListNode) -> ListNode:
    """
    Time: O(n * log(n))
    Space: O(1)
    """
    if not head or not head.next:
        return head
    mid = get_mid(head)
    left = sortList(head)
    right = sortList(mid)
    return merge(left, right)


def get_mid(head: ListNode) -> ListNode:
    mid_prev = head
    head = head.next
    while head and head.next:
        mid_prev = mid_prev.next
        head = head.next.next
    mid = mid_prev.next
    mid_prev.next = None
    return mid


def merge(first: ListNode, sec: ListNode) -> ListNode:
    if not first or not sec:
        return first or sec
    sentinel = ListNode()
    tail = sentinel
    while first and sec:
        if first.val < sec.val:
            tail.next = first
            first = first.next
        else:
            tail.next = sec
            sec = sec.next
        tail = tail.next
    tail.next = first if first else sec
    return sentinel.next
